src/branch_and_bound.rs

```diff
@@ -0,0 +1,582 @@
+        //      3
+        //
+        //
+        // 2_ value meets or exceeded target.
+        //    In this condition, we only backtrack one node
+        //
+        // From:
+        //      o
+        //     /
+        //    4
+        //   /
+        //  3
+        //
+        // To:
+        //      o
+        //     /
+        //    4
+        //   / \
+        //      3
+
+        // Set initial loop state
+        backtrack = false;
+        backtrack_subtree = false;
+
+        // * not enough value to make it to the target.
+        //   Therefore, explore a new new subtree.
+        if available_value + value < target {
+            backtrack_subtree = true;
+        }
+        // This optimization provides an upper bound on the amount of waste that is acceptable.
+        // Since value is lost when we create a change output due to increasing the size of the
+        // transaction by an output, we accept solutions that may be larger than the target as
+        // if they are exactly equal to the target and consider the overage waste or a throw
+        // away amount.  However we do not consider values greater than value + cost_of_change.
+        //
+        // This effectively creates a range of possible solution where;
+        // range = (target, target + cost_of_change]
+        //
+        // That is, the range includes solutions that exactly equal the target up to but not
+        // including values greater than target + cost_of_change.
+        else if value > target + cost_of_change {
+            backtrack = true;
+        }
+        // * value meets or exceeds the target.
+        //   Record the solution and the waste then continue.
+        //
+        // Check if index_selection is better than the previous known best, and
+        // update best_selection accordingly.
+        else if value >= target {
+            backtrack = true;
+
+            let current_waste = value - target;
+
+            if current_waste <= waste {
+                best_selection = Some(index_selection.clone());
+                waste = current_waste;
+            }
+        }
+
+        // * Backtrack
+        if backtrack {
+            let last_index = index_selection.pop().unwrap();
+            value -= utxo_candidate_amounts[last_index];
+            index -= 1;
+            assert_eq!(index, last_index);
+        }
+        // * Backtrack to new tree
+        else if backtrack_subtree {
+            // No new subtree left to explore.
+            if index_selection.is_empty() {
+                return index_to_utxo_list(best_selection, weighted_utxos);
+            }
+
+            // Anchor the new subtree at the next available index.
+            // The next iteration, the index will be incremented by one.
+            index = index_selection.remove(0);
+
+            // The available value of the next iteration.
+            available_value = utxo_candidate_amounts[index + 1..].iter().cloned().checked_sum()?;
```

Comment:
Doesn't matter much but `copied` would be more clear.

Reply:
`copied` is a minor degradation in performance in my testing.  Again, maybe switching to Criterion for better measurements should be a priority, but I show changing this from `cloned` to `copied` reduces performance from 830k ns/iter to 860k ns/iter.

Reply:
Something is seriously broken then. There's no good reason why those two should be different. Maybe something for bug report against rustc/llvm if you can minimize it.

Reply:
changed to Criterion and it shows less than a 2% variance after making this change (no noticeable change).